src/main/java/io/supertokens/webserver/api/bulkimport/BulkImportAPI.java

```diff
@@ -0,0 +1,202 @@
+            throw new ServletException(new BadRequestException("This API is not supported in the in-memory database."));
+        }
+
+        JsonObject input = InputParser.parseJsonObjectOrThrowError(req);
+        JsonArray users = InputParser.parseArrayOrThrowError(input, "users", false);
+
+        if (users.size() <= 0 || users.size() > BulkImport.MAX_USERS_TO_ADD) {
+            JsonObject errorResponseJson = new JsonObject();
+            String errorMsg = users.size() <= 0 ? "You need to add at least one user."
+                    : "You can only add " + BulkImport.MAX_USERS_TO_ADD + " users at a time.";
+            errorResponseJson.addProperty("error", errorMsg);
+            throw new ServletException(new WebserverAPI.BadRequestException(errorResponseJson.toString()));
+        }
+
+        AppIdentifier appIdentifier = null;
+        Storage storage = null;
+
+        try {
+            appIdentifier = getAppIdentifier(req);
+            storage = enforcePublicTenantAndGetPublicTenantStorage(req);
+        } catch (TenantOrAppNotFoundException | BadPermissionException e) {
+            throw new ServletException(e);
+        }
+
+        String[] allUserRoles = null;
+
+        try {
+            allUserRoles = StorageUtils.getUserRolesStorage(storage).getRoles(appIdentifier);
+        } catch (StorageQueryException e) {
+            throw new ServletException(e);
+        }
+
+        JsonArray errorsJson = new JsonArray();
+        List<BulkImportUser> usersToAdd = new ArrayList<>();
+
+        BulkImportUserUtils bulkImportUserUtils = new BulkImportUserUtils(allUserRoles);
+        for (int i = 0; i < users.size(); i++) {
+            try {
+                BulkImportUser user = bulkImportUserUtils.createBulkImportUserFromJSON(main, appIdentifier, users.get(i).getAsJsonObject(), Utils.getUUID());
+                usersToAdd.add(user);
+            } catch (io.supertokens.bulkimport.exceptions.InvalidBulkImportDataException e) {
+                JsonObject errorObj = new JsonObject();
+
+                JsonArray errors = e.errors.stream()
+                        .map(JsonPrimitive::new)
+                        .collect(JsonArray::new, JsonArray::add, JsonArray::addAll);
+
+                errorObj.addProperty("index", i);
+                errorObj.add("errors", errors);
+                errorsJson.add(errorObj);
+            } catch (StorageQueryException | TenantOrAppNotFoundException e) {
+                throw new ServletException(e);
+            }
+        }
+
+        if (errorsJson.size() > 0) {
+            JsonObject errorResponseJson = new JsonObject();
+            errorResponseJson.addProperty("error",
+                    "Data has missing or invalid fields. Please check the users field for more details.");
+            errorResponseJson.add("users", errorsJson);
+            throw new ServletException(new WebserverAPI.BadRequestException(errorResponseJson.toString()));
```

Comment:
what is an example output of this when several users have incorrect props in their JSON? How easy is is for the devs to know which user JSON entry to look at? Ideally, we should also have an index in the error pointing to exactly which input array's index has the issue

Reply:
See [this](https://github.com/supertokens/supertokens-core/pull/966#discussion_r1611557876)